### PR TITLE
Feature: Add the possibility to disable firewalls

### PR DESCRIPTION
**Pull request by delyriand**

### CODE REVIEW

File: src/EventListener/NoShopListener.php [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+namespace MonsieurBiz\SyliusNoCommercePlugin\EventListener;
+
+use MonsieurBiz\SyliusSettingsPlugin\Settings\SettingsInterface;
+use Sylius\Component\Channel\Context\ChannelContextInterface;
+use Symfony\Bundle\SecurityBundle\Security\FirewallMap;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Event\RequestEvent;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+
+final class NoShopListener
```
Comment (delyriand) [inline src/EventListener/NoShopListener.php]: Change the name of the listener, like `DisableFirewallListener`

File: src/EventListener/NoShopListener.php [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+{
+    /**
+     * @var FirewallMap
+     */
+    private $firewallContext;
```
Comment (delyriand) [inline src/EventListener/NoShopListener.php]: add type in class param definition

File: src/Form/Type/Settings/NoCommerceType.php [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    /**
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options): void
+    {
+        $choices = [];
+        foreach ($this->firewallRegistry as $firewall) {
+            $firewallConfig = $firewall->getConfig();
```
Comment (delyriand) [inline src/Form/Type/Settings/NoCommerceType.php:41]: exclude `admin` firewall to the choices list

File: src/Form/Type/Settings/NoCommerceType.php [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+            $firewallName = $firewallConfig->getName();
+            $choices[$firewallName] = $firewallConfig->getContext() ?? $firewallName;
+        }
+
+        $this->addWithDefaultCheckbox($builder, 'firewall_contexts', ChoiceType::class, [
+            'label' => 'Disable firewalls',
```
Comment (delyriand) [inline src/Form/Type/Settings/NoCommerceType.php]: add a translate label

File: src/Firewall/RegistryInterface.php [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+use Symfony\Bundle\SecurityBundle\Security\FirewallContext;
+
+interface RegistryInterface extends \IteratorAggregate
+{
+    public function addFirewall(FirewallContext $firewall): void;
+
+    /**
+     * @return FirewallContext[]
+     */
+    public function getIterator();
```
Comment (jacquesbh) [inline src/Firewall/RegistryInterface.php]: If the only need is the PHPDoc, you could add this doc at the top of the interface definition, and remove the duplicated declaration here.

File: src/Resources/config/services.yaml [hunk 6]
```diff
@@ -23,8 +23,23 @@ services:
...
+    Symfony\Bundle\SecurityBundle\Security\FirewallMap: '@security.firewall.map'
+
+    monsieurbiz.nocommerce.firewall.registry:
+        class: MonsieurBiz\SyliusNoCommercePlugin\Firewall\Registry
+
+    MonsieurBiz\SyliusNoCommercePlugin\Firewall\RegistryInterface: '@monsieurbiz.nocommerce.firewall.registry'
+    
+    MonsieurBiz\SyliusNoCommercePlugin\Form\Type\Settings\NoCommerceType:
+        arguments:
+            - '@monsieurbiz.nocommerce.firewall.registry'
```
Comment (jacquesbh) [inline src/Resources/config/services.yaml]: This is useless if you use autowiring, which is the case by default for all settings I think.

File: src/Resources/config/monsieurbiz/settings.yaml [hunk 7]
```diff
@@ -0,0 +1,12 @@
+monsieurbiz_sylius_settings:
+    plugins:
+        monsieurbiz.nocommerce:
+            vendor_name: Monsieur Biz
+            vendor_url:
+            plugin_name: No Commerce
+            description: No Commerce configuration
+            icon: shopping bag
+            use_locales: false
```
Comment (jacquesbh) [inline src/Resources/config/monsieurbiz/settings.yaml]:
```suggestion
            use_locales: true
```

We could accept to have a shop only for a specific locale, right?

File: src/Firewall/Registry.php [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+namespace MonsieurBiz\SyliusNoCommercePlugin\Firewall;
+
+use Symfony\Bundle\SecurityBundle\Security\FirewallContext;
+
+class Registry implements RegistryInterface
+{
+    /**
+     * @var FirewallContext[]
+     */
+    private array $firewalls;
```
Comment (jacquesbh) [inline src/Firewall/Registry.php]:
```suggestion
    private array $firewalls = [];
```

File: src/Form/Type/Settings/NoCommerceType.php [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    public function buildForm(FormBuilderInterface $builder, array $options): void
+    {
+        $choices = [];
+        foreach ($this->firewallRegistry as $firewall) {
+            $firewallConfig = $firewall->getConfig();
+            $firewallName = $firewallConfig->getName();
+            $choices[$firewallName] = $firewallConfig->getContext() ?? $firewallName;
+        }
+
+        $this->addWithDefaultCheckbox($builder, 'firewall_contexts', ChoiceType::class, [
```
Comment (jacquesbh) [inline src/Form/Type/Settings/NoCommerceType.php]:
```suggestion
        $this->addWithDefaultCheckbox($builder, 'disabled_firewall_contexts', ChoiceType::class, [
```

**Review by jacquesbh — CHANGES_REQUESTED** [review]

File: src/Resources/config/monsieurbiz/settings.yaml (hunk 7 above)
Comment (delyriand) [reply src/Resources/config/monsieurbiz/settings.yaml]: yes, it's possible to want disable firewalls only for a locale

File: src/Resources/config/services.yaml (hunk 6 above)
Comment (delyriand) [reply src/Resources/config/services.yaml]: i can't remove this without an error

File: composer.json [hunk 10]
```diff
@@ -7,7 +7,8 @@
     "require": {
         "php": "~7.4",
         "sylius/sylius": "^1.8",
-        "phpstan/phpstan-webmozart-assert": "^0.12.7"
+        "phpstan/phpstan-webmozart-assert": "^0.12.7",
+        "monsieurbiz/sylius-settings-plugin": "^1.0.0"
```
Comment (jacquesbh) [inline composer.json:11]: You need to install the plugin in the test application.

File: src/EventListener/DisableFirewallListener.php [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+        $this->channelContext = $channelContext;
+    }
+
+    public function __invoke(RequestEvent $event): void
+    {
+        $currentChannel = $this->channelContext->getChannel();
+        $disabledFirewall = $this->nocommerceSettings->getCurrentValue($currentChannel, null, 'disabled_firewall_contexts') ?? [];
+        $firewallContextName = $this->getFirewallContextName($event->getRequest());
+        if (\in_array($firewallContextName, $disabledFirewall, true)) {
+            throw new NotFoundHttpException('Route not found');
```
Comment (jacquesbh) [inline src/EventListener/DisableFirewallListener.php]: I don't know why but this exception finishes in a 500 HTTP error.

We should have a 404.

**Review by jacquesbh — CHANGES_REQUESTED** [review]

File: src/EventListener/DisableFirewallListener.php (hunk 11 above)
Comment (delyriand) [reply src/EventListener/DisableFirewallListener.php]: @jacquesbh i added a check, i'm checking that it's a master request: https://github.com/monsieurbiz/SyliusNoCommercePlugin/pull/13/files#diff-ca2600a1ed3414104f0833ffa2751a07ae28b35350ecb0fd14b5e307037f79bbR41-R48

(and i excluded the profiler routes from my firewall check)

File: composer.json (hunk 10 above)
Comment (delyriand) [reply composer.json:11]: i create a new PR to update the test application, based on your WIP PR

**Review by jacquesbh — APPROVED** [review]

File: src/EventListener/DisableFirewallListener.php [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+        $this->channelContext = $channelContext;
+    }
+
+    public function __invoke(RequestEvent $event): void
+    {
+        if (!$event->isMasterRequest()) {
+            return;
+        }
+        $request = $event->getRequest();
+        if (\in_array($request->attributes->get('_route'), ['_wdt', '_profiler', '_profiler_search', '_profiler_search_results'], true)) {
```
Comment (lanfisis) [inline src/EventListener/DisableFirewallListener.php]: Values `['_wdt', '_profiler', '_profiler_search', '_profiler_search_results']` in a class constant maybe?

File: src/EventListener/DisableFirewallListener.php [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    private function getFirewallContextName(Request $request): string
+    {
+        $firewallConfig = $this->firewallContext->getFirewallConfig($request);
+        if (null === $firewallConfig) {
+            return '';
+        }
+
+        return $firewallConfig->getContext() ?? $firewallConfig->getName();
```
Comment (lanfisis) [inline src/EventListener/DisableFirewallListener.php:72]: 🤨  ?

**Review by lanfisis — CHANGES_REQUESTED** [review]

File: src/EventListener/DisableFirewallListener.php (hunk 13 above)
Comment (delyriand) [reply src/EventListener/DisableFirewallListener.php:72]: a firewall may not have context code, so i using the context name as fallback

**Review by lanfisis — APPROVED** [review]

**Comment by lanfisis** [comment]
Thanks a lot for this 👍

**Pull request merged**